[AMDGPU][SIInsertWaitcnt] Implement Waitcnt Expansion for Profiling

PankajDwivedi-25 opened the pull request:
Reference issue: https://github.com/ROCm/llvm-project/issues/67

This patch adds support for expanding s_waitcnt instructions into sequences with decreasing counter values, enabling PC-sampling profilers to identify which specific memory operation is causing a stall.

This is controlled via:
Clang flag: -mamdgpu-expand-waitcnt-profiling / -mno-amdgpu-expand-waitcnt-profiling
Function attribute: "amdgpu-expand-waitcnt-profiling"

When enabled, instead of emitting a single waitcnt, the pass generates a sequence that waits for each outstanding operation individually. For example, if there are 5 outstanding memory operations and the target is to wait until 2 remain:


**Original**: 
s_waitcnt vmcnt(2)

**Expanded**:  
s_waitcnt vmcnt(4)
s_waitcnt vmcnt(3)
s_waitcnt vmcnt(2)

The expansion starts from (Outstanding - 1) down to the target value, since waitcnt(Outstanding) would be a no-op (the counter is already at that value).

- Uses ScoreBrackets to determine the actual number of outstanding operations
- Only expands when operations complete in-order
- Skips expansion for mixed event types (e.g., LDS+SMEM on same counter)
- Skips expansion for scalar memory (always out-of-order)

Releated previous work for Reference
- **PR**: llvm/llvm-project#79236 (related `-amdgpu-waitcnt-forcezero`)

jayfoad commented:
> > Why would you restrict this to "non-zero counter values"?
> 
> When a waitcnt already has a zero counter value expanding it would just generate another waitcnt(0), which provides no additional profiling granularity. If you believe there's a use case for expanding waitcnt(0), I'd be happy to discuss it.

The requirement is that instead of emitting e.g. `s_waitcnt vmcnt(2)` you should emit e.g.:
```
s_waitcnt vmcnt(4)
s_waitcnt vmcnt(3)
s_waitcnt vmcnt(2)
```
The starting value "4" here is assuming that SIInsertWaitcnts already knows that the upper bound on this counter's value is 5, so 4 is the highest value you can wait for that will have any effect.

Similarly instead of `s_waitcnt vmcnt(0)` you should emit:
```
s_waitcnt vmcnt(4)
s_waitcnt vmcnt(3)
s_waitcnt vmcnt(2)
s_waitcnt vmcnt(1)
s_waitcnt vmcnt(0)
```

> > Why does this need a new subtarget feature?
> 
> On this, I am still not fully sure what would be the best approach to handle. I there any suggestion from you?

I don't know why you added a subtarget feature. I would suggest not adding it -- unless there is some reason for it that I am missing.

arsenm commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
Use of subtarget feature still here, this needs to be replaced with something else (maybe an attribute)

PankajDwivedi-25 commented:
> > > Why would you restrict this to "non-zero counter values"?
> > 
> > 
> > When a waitcnt already has a zero counter value expanding it would just generate another waitcnt(0), which provides no additional profiling granularity. If you believe there's a use case for expanding waitcnt(0), I'd be happy to discuss it.
> 
> The requirement is that instead of emitting e.g. `s_waitcnt vmcnt(2)` you should emit e.g.:
> 
> ```
> s_waitcnt vmcnt(4)
> s_waitcnt vmcnt(3)
> s_waitcnt vmcnt(2)
> ```
> 
> The starting value "4" here is assuming that SIInsertWaitcnts already knows that the upper bound on this counter's value is 5, so 4 is the highest value you can wait for that will have any effect.
> 
> Similarly instead of `s_waitcnt vmcnt(0)` you should emit:
> 
> ```
> s_waitcnt vmcnt(4)
> s_waitcnt vmcnt(3)
> s_waitcnt vmcnt(2)
> s_waitcnt vmcnt(1)
> s_waitcnt vmcnt(0)
> ```

looks like outstanding value is always 0 in these cases.

PankajDwivedi-25 committed:
[AMDGPU] Add -amdgpu-expand-waitcnt-profiling option for PC-sampling profiling

PankajDwivedi-25 committed:
add run line for diff GPU Gen and counter types

PankajDwivedi-25 committed:
Address reviewer feedback: fix getWaitCountMax and reduce code duplication

- Fix getWaitCountMax() to use correct bitmasks based on architecture:
  - Pre-GFX12: Use getVmcntBitMask/getLgkmcntBitMask for LOAD_CNT/DS_CNT
  - GFX12+: Use getLoadcntBitMask/getDscntBitMask for LOAD_CNT/DS_CNT
- Refactor repetitive if-blocks for LOAD_CNT, DS_CNT, EXP_CNT into
  a single loop using getCounterRef helper function
- Fix X_CNT to return proper getXcntBitMask(IV) instead of 0

PankajDwivedi-25 commented:
> I'm still not sure how this works with out of order counters.

I see the point, you mean the event which share the same counter might finish out-of-order right? this will cause issue because I haven't made any check here. 

Is it problematic or it is still useful for profiling purposes?

One quick fix that arise in my mind is skip expanding out of order events. exploring if we can also check if they are really different events rather than just returning true if more than one bit is set. @Pierre-vh

PankajDwivedi-25 committed:
skip expanding out-of-order events

PankajDwivedi-25 committed:
fix: resolve issue after rebase

PankajDwivedi-25 committed:
add more test

Pierre-vh commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -1527,38 +1564,108 @@ bool WaitcntGeneratorPreGFX12::applyPreexistingWaitcnt(
+  // Helper to emit expanded waitcnt sequence for profiling.
+  // Emits waitcnts from (Outstanding-1) down to Target, or just Target if
+  // nothing to expand. The EmitWaitcnt callback emits a single waitcnt.
+  auto emitExpandedWaitcnt = [&](unsigned Outstanding, unsigned Target,
```
```suggestion
  auto EmitExpandedWaitcnt = [&](unsigned Outstanding, unsigned Target,
```

PankajDwivedi-25 committed:
Update SIInsertWaitcnts.cpp

Co-authored-by: Pierre van Houtryve <pierre.vanhoutryve@amd.com>

PankajDwivedi-25 committed:
Update SIInsertWaitcnts.cpp

Co-authored-by: Pierre van Houtryve <pierre.vanhoutryve@amd.com>

PankajDwivedi-25 committed:
review: move hardwareLimit inside AMDGPUBaseInfo

PankajDwivedi-25 committed:
review: use function attr instead cl::opt flag

Pierre-vh reviewed (approved):
Implementation now LGTM, but please wait a bit for other reviewers to approve as well as I wasn't the only one with comments

PankajDwivedi-25 committed:
avoid duplicating getWaitCountMax
